### DISCUSSION TITLE
Trailer features

### DISCUSSION THREAD

**Discussion by Kishore1406** [jpos/jPOS#534] [General]
Hi 

We faced an issue in adding the trailer to the X25 channel, so we worked on JPOS 2.1.8 and added the trailer to the X25 channel.

While working on it, we found that you deprecated the **_sendMessageTrailer_** in the Base channel. May we know in the future if there is any plan for adding trailer, or will we give a solution to add trailers to the Base channel?

**Comment by alcarraz** [comment]
You can use the correctly spelled version [sendMessageTrailer](http://jpos.org/doc/javadoc/org/jpos/iso/BaseChannel.html#sendMessageTrailer-org.jpos.iso.ISOMsg-byte:A-)

**Comment by Kishore1406** [comment]
We altered the [sendMessageTrailer](http://jpos.org/doc/javadoc/org/jpos/iso/BaseChannel.html#sendMessageTrailer-org.jpos.iso.ISOMsg-byte:A-) to pass the trailer in the x25 channel; it was working, and we added this solution in the basechannel. The trailer will work for all channels.

Existing Code:
![MicrosoftTeams-image (3)](https://github.com/jpos/jPOS/assets/102660941/91100272-9be8-4def-8e61-fef209342f00)

Customized Code:
<img width="524" alt="MicrosoftTeams-image (4)" src="https://github.com/jpos/jPOS/assets/102660941/b5eb8368-03b6-45f8-8907-c55d3b1d3d0b">

Please provide the guidelines for merge our code into JPOS source code.

**Reply by ar** [reply]
The trailer thing is legacy stuff that comes from days where the technical people at financial institutions didn't believe TCP/IP was indeed a reliable transport, and the whole internet runs on top of it, so they invented things such as message headers and trailers (STX...ETX). I believe we are OK as we are now in BaseChannel, if somebody needs to implement trailers, she/he can extend BaseChannel.

**Reply by alcarraz** [reply]
@Kishore1406 have you tried [`BASE24TCPChannel`](https://github.com/jpos/jPOS/blob/master/jpos/src/main/java/org/jpos/iso/channel/BASE24TCPChannel.java) or [`BASE24Channel`](https://github.com/jpos/jPOS/blob/master/jpos/src/main/java/org/jpos/iso/channel/BASE24Channel.java)?

What is the trailer you need to send? Or you can just extend `X25Channel` and send and receive the trailer the same way `Base24Channel` does.

**Reply by Kishore1406** [reply]
@alcarraz 
We tried [BASE24TCPChannel](https://github.com/jpos/jPOS/blob/master/jpos/src/main/java/org/jpos/iso/channel/BASE24TCPChannel.java) and [BASE24Channel](https://github.com/jpos/jPOS/blob/master/jpos/src/main/java/org/jpos/iso/channel/BASE24Channel.java) but it can add only 03(ETX) has trailer. We want **03d858** has trailer so we override the source code in BASEChannel to pass the dynamic trailer for all channels. Our business scenario X25 channel only receiving the response.

**Reply by alcarraz** [reply]
But why not extending one of them and add those bytes, taking Base24 or Base24TCP as a reference?

**Reply by Kishore1406** [reply]
But if we are extending any one of those channels [BASE24TCPChannel](https://github.com/jpos/jPOS/blob/master/jpos/src/main/java/org/jpos/iso/channel/BASE24TCPChannel.java) and [BASE24Channel](https://github.com/jpos/jPOS/blob/master/jpos/src/main/java/org/jpos/iso/channel/BASE24Channel.java) it will be specific channel now we have add it in base channel so that it can be applicable to all the channel

**Reply by alcarraz** [reply]
Thing is, most of the channels, if not all, that don't override `sendTrailer` will just fail when a message with a trailer is sent over them with this modification.

**Reply by alcarraz** [reply]
Maybe, one option, could be to make it configurable in a boolean property, if the channel should honor the message trailer, but to make things consistent, you should also set the trailer on the message on receive.

This will take more work, and should be highly tested before merging it. It would touch many sensible places of the `BaseChannel` behavior, which is also at the core of the jPOS project.

**Comment by it240884sii** [comment]
it`s time to remove incorrectly spelled version and annotation @SuppressWarnings("deprecation") from correctly spelled. ready or not here i come =)

**Reply by ar** [reply]
We'll certainly do that in jPOS 3.0 ('next' branch).

**Reply by it240884sii** [reply]
life is to short for waiting next release...

**Reply by ar** [reply]
Not sure what's your point. The properly spelled methods are already available, and the old ones are flagged for deprecation.
